source/log.c

```diff
@@ -0,0 +1,386 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#ifndef _CRT_SECURE_NO_WARNINGS
+	#define _CRT_SECURE_NO_WARNINGS
+#endif
+
+#ifndef _CRT_NONSTDC_NO_DEPRECATE
+	#define _CRT_NONSTDC_NO_DEPRECATE
+#endif
+
+#include <aws/common/log.h>
+
+#include <aws/common/atomic.h>
+#include <aws/common/common.h>
+#include <aws/common/memory_pool.h>
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+
+
+#include <stdio.h>
+#include <time.h>
+#include <inttypes.h>
+
+#define AWS_SINGLY_LIST_REVERSE(T, list) \

```

Comment:
Originally I was not sure how many different typed linked lists there were going to be. If you peek inside the macro a temporary pointer is created, and the type is needed to do this. However, since it was only used for one single pointer type, it could easily be an inline function instead. In the end I didn't think it was important enough to justify spending time changing it to an inline function.

Reply:
We can do something similar to doubly-linked lists and define a

```
struct aws_single_list_node {
  struct aws_single_list_node *next;
};

void aws_single_list_reverse(struct aws_single_list_node *head);
```
AWS_CONTAINER_OF (in linked_list.h) can be used to get back to the containing structure.